sparse_strips/vello_api/src/paths.rs

```diff
@@ -0,0 +1,130 @@
+// Copyright 2025 the Vello Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! A collection of vector graphics paths, for use in.
+//!
+//! This is an internal implementation detail of [`Scene`](crate::Scene).
+//! If you're a consumer of Vello API writing an application, you do not need to use this API.
+//! Instead, this is exposed for use by [`Renderer`](crate::Renderer)s.
+//!
+//! This design has been made with "path caching" in mind, which will allow using rasterised forms of
+//! paths (such as glyphs) directly in the renderer, instead of rasterising from scratch each frame.
+//! This can massively improve performance and efficiency on subsequent frames.
+//! This is however not implemented by any backend, so isn't provided for in Vello API.
+//! It also ensures that there are very few per-frame allocations (i.e. avoids allocating
+//! for each path).
+
+use alloc::vec::Vec;
+use peniko::{
+    Fill,
+    kurbo::{PathEl, Shape, Stroke},
+};
+
+#[derive(Debug, Clone, Copy)]
+/// The id for a single path within a given [`PathSet`].
+/// This is an index into the [`meta`](PathSet::meta) field.
+// In a future world with path caching, this would be paired with a path group id.
+// For "scene-local" paths, you would then use a marker "local" path group id.
+pub struct PathId(pub u32);
+
+#[derive(Debug)]
+pub struct PathSet {
+    // There are arguments for a "dynamic length" encoding here, as PathEl is sized for 6 f64s (plus a disciminant)
+    // It depends somewhat on what proportion of the elements are a CurveTo
+    pub elements: Vec<PathEl>,
+    pub meta: Vec<PathMeta>,
+}
+
+impl PathSet {
+    pub fn new() -> Self {
+        Self {
+            elements: Vec::new(),
+            meta: Vec::new(),
+        }
+    }
+
+    pub fn clear(&mut self) {
+        self.elements.clear();
+        self.meta.clear();
+    }
+}
+
+impl Default for PathSet {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PathSet {
+    pub fn prepare_fill(&mut self, fill_rule: Fill, shape: &impl Shape) -> PathId {
+        let start_index = self.elements.len();
+        // TODO: Maybe determine the optimal tolerance from the transform.
+        // If this can be post-transformed, that doesn't actually meaningfully apply?
+        self.elements.extend(shape.path_elements(0.1));
+        let meta_index = self.meta.len();
+        self.meta.push(PathMeta {
+            start_index,
+            operation: Operation::Fill(fill_rule),
+        });
+
+        // TODO: Better error handling here?
+        PathId(meta_index.try_into().unwrap())
+    }
+
+    // TODO: Maybe require stroke expansion happen *prior* to this stage?
+    pub fn prepare_stroke(&mut self, stroke_rule: Stroke, shape: &impl Shape) -> PathId {
+        let start_index = self.elements.len();
+        // TODO: Determine the optimal tolerance from the transform.
+        // TODO: Perform dash expansion now, even if not stroke expansion?
```

Comment:
We should create an issue, or at least have a discussion of the trade-offs involved somewhere.